docs/streaming-programming-guide.md

```diff
@@ -630,35 +630,106 @@ which creates a DStream from text
+{% endhighlight %}
+For text files
+
+{% highlight scala %}
+streamingContext.textFileStream(dataDirectory)
+{% endhighlight %}
+</div>
 
-	For simple text files, there is an easier method `streamingContext.textFileStream(dataDirectory)`. And file streams do not require running a receiver, hence does not require allocating cores.
+<div data-lang="java" markdown="1">
+{% highlight java %}
+streamingContext.fileStream<KeyClass, ValueClass, InputFormatClass>(dataDirectory);
+{% endhighlight %}
+For text files
 
-	<span class="badge" style="background-color: grey">Python API</span> `fileStream` is not available in the Python API, only	`textFileStream` is	available.
+{% highlight java %}
+streamingContext.textFileStream(dataDirectory);
+{% endhighlight %}
+</div>
 
-- **Streams based on Custom Receivers:** DStreams can be created with data streams received through custom receivers. See the [Custom Receiver
+<div data-lang="python" markdown="1">
+`fileStream` is not available in the Python API; only `textFileStream` is available.
+{% highlight python %}
+streamingContext.textFileStream(dataDirectory)
+{% endhighlight %}
+</div>
+
+</div>
+
+##### How Directories are Monitored
+{:.no_toc}
+
+Spark Streaming will monitor the directory `dataDirectory` and process any files created in that directory.
+
+   * A simple directory can be monitored, such as `hdfs://namenode:8040/logs/`.
+     All files directly under such a path will be processed as they are discovered.
+   + A [POSIX glob pattern](http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_13_02) can be supplied, such as
+     `hdfs://namenode:8040/logs/2017/*`.
+     Here, the DStream will consist of all files in the directories
+     matching the pattern.
+     That is: it is a pattern of directories, not of files in directories.
+   + All files must be in the same data format.
+   * A file is considered part of a time period based on its modification time,
+     not its creation time.
+   + Once processed, changes to a file within the current window will not cause the file to be reread.
+     That is: *updates are ignored*.
+   + The more files under a directory, the longer it will take to
+     scan for changes —even if no files have been modified.
+   * If a wildcard is used to identify directories, such as `hdfs://namenode:8040/logs/2016-*`,
+     renaming an entire directory to match the path will add the directory to the list of
+     monitored directories. Only the files in the directory whose modification time is
+     within the current window will be included in the stream.
+   + Calling `FileSystem.setTimes()` to fix the timestamp is a way to have the file picked
+     up in a later window, even if its contents have not changed.
+
+
+##### Streaming to FileSystems vs Object stores
+{:.no_toc}
+
+"Full" Filesystems such as HDFS tend to set the modification time on their files as soon
```

Comment:
Is the message of this section: 

Write new data to a file outside the monitored directory, and then move it into that directory after it's closed. Otherwise, if writing directly to a file in the monitored directory, it may be processed while it's being written to. However: if you are working with an object store like S3, this approach still works correctly but may be slow because a rename can be a copy. In some cases it's OK to write directly into the destination directory because the object store won't expose the file until it's done.

If so I think this could be a little more direct, telling users what they need to know to use this correctly. It seems like renaming is always at least correct. I think that if different advice applies for S3, that should also be called out by name, or else I'm left wondering unnecessarily whether this applies to the most commonly-used store, S3.